### PR TITLE
Allow plugins to define their own targets

### PR DESCRIPTION
**Pull request by AdrienClairembault**
Allow plugin to add additional targets.
This is done through a `add_formcreator_targets` hook like this:

```php
$PLUGIN_HOOKS['add_formcreator_targets']['plugin_demo'] = [
      EmailTarget::class
];
```

### CODE REVIEW

File: inc/hooksmigration.class.php
Comment (btry) [inline inc/hooksmigration.class.php:51]: This create table statmeent must be in the plugin_formcreator_empty.sql file with a IF NOT EXIST condition.

This way, upgrade will handle it.

**Comment by AdrienClairembault** [comment]
Rebased and updated the description.
Should be OK to review now :)

File: inc/hooksmigration.class.php
Comment (AdrienClairembault) [reply inc/hooksmigration.class.php:51]: No tables were added, this is only an helper method to be used by others plugins.
This allows them to easily create the 'base' target table in one line of code.

They can then add their specific fields manually.

**Comment by AdrienClairembault** [comment]
Up

**Review by btry — CHANGES_REQUESTED** [review]

File: inc/abstracttarget.class.php
Comment (AdrienClairembault) [reply inc/abstracttarget.class.php:2607]: I need to use this method in my plugin.

```php
public static function displayTabContentForItem(CommonGLPI $item, $tabnum = 1, $withtemplate = 0) {
    switch ($item->getType()) {
        case self::class:
            /** @var EmailTarget $item */
            switch ($tabnum) {
                case 1:
                    self::showConditions($item);
                    break;
            }
            break;
    }

    return false;
}
```

How will it be done with twig ?

File: inc/abstracttarget.class.php
Comment (AdrienClairembault) [reply inc/abstracttarget.class.php:300]: How can I share this code with an item that is a `PluginFormcreatorTargetInterface` but not an 
`AbstractTarget ` ?

File: inc/form.class.php
Comment (AdrienClairembault) [reply inc/form.class.php:1993]: Indeed, I wanted to rename it but forgot.
Done b7022860.

File: inc/abstracttarget.class.php
Comment (btry) [reply inc/abstracttarget.class.php:300]: If you have such need we may solve it by creating an intermediate abstract class. Something like this.

CommonDBTM
|- PluginFormcreatorAbstractTarget implements PluginFormcreatorTargetInterface
    |- PluginFormcreatorAbstractItilTarget
    |   |- PluginFormcreatorTargetTicket
    |   |- PluginFormcreatorTargetChange
    |   |- PluginFormcreatorTargetProblem
    |- PluginFooTargetBar

We will probably need to split the methods of PluginFormcreatorAbstractTarget to move some in PluginFormcreatorAbstractItilTarget.

File: inc/abstracttarget.class.php
Comment (btry) [reply inc/abstracttarget.class.php:2607]: Like this

```twig
                    {% set conditions = call('PluginFormcreatorCondition::getConditionsFromItem', [item]) %}
                    {% for condition in conditions %}
                        {{ include('@formcreator/components/form/condition.html.twig') }}
                    {% endfor %}
```

File: ajax/condition.php
Comment (AdrienClairembault) [reply ajax/condition.php]: No longer relevant, this code was removed since namespace were fixed in another core PR.

**Comment by AdrienClairembault** [comment]
I've rebased on #2647 to do some testing (so the diff is a bit broken for now).
Will be all good once #2647 is merged and I can rebase on develop

**Comment by btry** [comment]
Hi

#2647 : unit tests are green; it is waiting for your review.

**Comment by AdrienClairembault** [comment]
> Hi
> 
> #2647 : unit tests are green; it is waiting for your review.

Review was done this morning, you may have missed it ;)

File: inc/form.class.php
Comment (AdrienClairembault) [inline inc/form.class.php:464]: Seems like this token wasn't being used.

**Comment by AdrienClairembault** [comment]
Ready for review :)

File: js/scripts.js
Comment (btry) [inline js/scripts.js:1329]: It would be better to keep this as a function, but moving it in plugin_formcreator.

**Review by btry — APPROVED** [review]

**Pull request merged**
